agent/grpc-middleware/rate.go

```diff
@@ -0,0 +1,55 @@
+package middleware
+
+import (
+	"context"
+	"errors"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/peer"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/tap"
+
+	"github.com/hashicorp/consul/agent/consul/rate"
+)
+
+// ServerRateLimiterMiddleware implements a ServerInHandle function to perform
+// RPC rate limiting at the cheapest possible point (before the full request has
+// been decoded).
+func ServerRateLimiterMiddleware(limiter RateLimiter) tap.ServerInHandle {
+	return func(ctx context.Context, info *tap.Info) (context.Context, error) {
+		peer, ok := peer.FromContext(ctx)
+		if !ok {
+			// This should never happen!
+			return ctx, status.Error(codes.Internal, "gRPC rate limit middleware unable to read peer")
+		}
+
+		err := limiter.Allow(rate.Operation{
+			Name:       info.FullMethodName,
+			SourceAddr: peer.Addr,
+			// TODO: operation type.
+		})
+
+		switch {
+		case err == nil:
+			return ctx, nil
+		case errors.Is(err, rate.ErrRetryElsewhere):
+			return ctx, status.Error(codes.ResourceExhausted, err.Error())
+		default:
```

Comment:
Great call! I think it'd be ok to fail with `codes.Internal` for any unknown error, as its a bug in Consul that should be surfaced during our testing.

You raise a good point about `nil` pointers though, as this will presumably be called before/in a different code-path to our recovery interceptor so we probably need to handle panics itself.